File: leetcode/easy/27-removeElement.js
// Given an array nums and a value val, remove all instances of that value in-place and return the new length.

// Do not allocate extra space for another array, you must do this by modifying the input array in-place with O(1) extra memory.

// The order of elements can be changed. It doesn't matter what you leave beyond the new length.

// Clarification:

// Confused why the returned value is an integer but your answer is an array?

// Note that the input array is passed in by reference, which means a modification to the input array will be known to the caller as well.

var removeElement = function(nums, val) {
    if(nums.length === 0) return 0;
    if(nums.length === 1 && nums[0] === val) return 0;
    
    let i = 0;
    
    while(i < nums.length) {
        if(nums[i] === val) {
            nums.splice(i, 1);
        } else {
            i++;
        }
    }
    return nums.length;
};

File: leetcode/easy/27-removeElement.cpp
class Solution {
public:
    int removeElement(vector<int>& nums, int val) {
        int i = 0;

        for (; i < nums.size(); i++) {
            if (nums[i] == val) {
                int j = i + 1;
                while (j < nums.size() && nums[j] == val) { j++; }
                if (j >= nums.size() || nums[j] == val) break;
                swap(nums[i], nums[j]);
            }
        }
        return i;
    }
};
